src/main/scala/se/nimsa/dcm4che/streams/DicomFlows.scala

```diff
@@ -62,34 +70,167 @@ object DicomFlows {
             push(out, chunk)
           else {
             buffer = buffer ++ chunk
-            if (buffer.length >= 140)
-              if (isPreamble(buffer))
-                if (isHeader(buffer.drop(132)))
-                  setValidated()
-                else
+
+            if (buffer.length >= maxBufferLength) {
+              if (isPreamble(buffer)) {
+                if (isHeader(buffer.drop(132))) {
+                  if (contexts.isDefined) {
+                    val info = DicomParsing.dicomInfo(buffer.drop(132)).get
+                    validateFileMetaInformation(buffer.drop(132), info)
+                  } else {
+                    setValidated()
+                  }
+                } else {
                   setFailed()
-              else if (isHeader(buffer))
-                setValidated()
-              else
+                }
+              } else if (isHeader(buffer)) {
+                if (contexts.isDefined) {
+                  val info = DicomParsing.dicomInfo(buffer).get
+                  validateSOPClassUID(buffer, info)
+                } else {
+                  setValidated()
+                }
+              } else {
                 setFailed()
-            else
+              }
+            } else {
               pull(in)
+            }
           }
         }
 
         override def onUpstreamFinish() = {
           if (!isValidated)
-            if (buffer.length == 132 && isPreamble(buffer))
-              setValidated()
-            else if (buffer.length >= 8 && isHeader(buffer))
-              setValidated()
-            else
-              setFailed()
+            if (contexts.isDefined) {
+              if (buffer.length >= 132 && isPreamble(buffer)) {
+                val info = DicomParsing.dicomInfo(buffer.drop(132)).get
+                validateFileMetaInformation(buffer.drop(132), info)
+              } else if (buffer.length >= 8 && isHeader(buffer)) {
+                val info = DicomParsing.dicomInfo(buffer).get
+                validateSOPClassUID(buffer, info)
+              } else {
+                setFailed()
+              }
+            } else {
+              if (buffer.length == 132 && isPreamble(buffer))
+                setValidated()
+              else if (buffer.length >= 8 && isHeader(buffer))
+                setValidated()
+              else
+                setFailed()
+            }
           completeStage()
         }
 
         def isHeader(data: ByteString) = DicomParsing.dicomInfo(data).isDefined
 
+        def readHeader(buffer: ByteString, assumeBigEndian: Boolean, explicitVR: Boolean): (Int, VR, Int, Int) = {
```

Comment:
This method overlaps alot with `DicomPartFlow.readHeader`. Is there anything that perhaps can be generalized and moved to e.g. `DicomParsing`?